README.md
# rdrct

rdrct is a URL shortener and redirection service.

## Quick start

To run rdrct, the simplest option is to do so using Docker. For that, first you have to build the Docker image:

```shell
$ docker build -t rdrct .
```

Once the build has been successfully completed, run the Docker image using the following command. Please note that you have to provide some values for the environment variables `API_USERNAME` and `API_PASSWORD`:

```shell
$ docker run \
    -d \
    -p 3000:3000 \
    -e API_USERNAME=<username> \
    -e API_PASSWORD=<password> \
    --init \
    --name rdrct \
    rdrct
```

The most common interaction is following a redirect. For that, send a `GET` request to rdrct and provide the `key` of the desired redirect as path, e.g.:

```
GET http://localhost:3000/tnw
```

Depending on how the redirect has been configured, you will either be redirected permanently (`301`) or temporarily (`307`). If the key you provide does not exist, you get a `404`.

### Interacting with the API

For managing the redirects, rdrct provides an API. To connect to this API you always have to send the username and the password you set when you started the Docker image using the `API_USERNAME` and `API_PASSWORD` environment variables.

*Please note that rdrct uses HTTP basic authentication, which does not encrypt the credentials in any way. This will leak your credentials to anyone who is able to intercept the traffic between your client and the server. While this is fine for running things locally while developing or trying things out, never run it like this in production. Make sure to always run it behind a reverse proxy that acts as a TLS endpoint in order to run things securely!*

### Getting all redirects

To get a list of all existing redirects, send a `GET` request to the `/api/redirects` route.

### Adding redirects

To add a new redirect, send a `POST` request to the `/api/add-redirect` route, and provide the data for the desired redirect in the request body. The request body must match the following JSON schema:

```json
{
  "type": "object",
  "properties": {
    "key": { "type": "string", "minLength": 1 },
    "url": { "type": "string", "minLength": 1, "format": "uri" },
    "type": { "type": "string", "enum": [ "permanent", "temporary" ]}
  },
  "required": [ "key", "url", "type" ],
  "additionalProperties": false
}
```

If the redirect was added successfully, you get back a `200`. If it already existed, you will get a `409`.

### Editing redirects

To edit a redirect, send a `POST` request to the `/api/edit-redirect` route, and provide the new URL for the desired redirect in the request body. The request body must match the following JSON schema:

```json
{
  "type": "object",
  "properties": {
    "key": { "type": "string", "minLength": 1 },
    "url": { "type": "string", "minLength": 1, "format": "uri" }
  },
  "required": [ "key", "url" ],
  "additionalProperties": false
}
```

If the redirect was edited successfully, you get back a `200`. If it could not be found, you will get a `404`.

### Removing redirects

To remove a redirect, send a `POST` request to the `/api/remove-redirect` route, and provide the key of the redirect in the request body. The request body must match the following JSON schema:

```json
{
  "type": "object",
  "properties": {
    "key": { "type": "string", "minLength": 1 }
  },
  "required": [ "key" ],
  "additionalProperties": false
}
```

If the redirect was removed successfully, you get back a `200`. If the redirect could not be found, you will get a `404`.

*Please note that removing a redirect also removes all of its statistics.*

### Getting statistics for a redirect

To get statistics on how often a redirect has been used, send a `GET` request to the `/api/statistics/<key>` route and replace `<key>` with the key of the desired redirect.

Maybe you do not want to fetch all-time statistics, but only for a limited time frame. For that, you can specify an optional `from` and / or `to` parameter using the query string to hand over a timestamp.

If the statistics could be fetched, you get back a `200`. If the redirect could not be found, you will get a `404`.

### Getting statistics for all redirects

To get statistics on how often all redirects have been used, send a `GET` request to the `/api/statistics` route.

Maybe you do not want to fetch all-time statistics, but only for a limited time frame. For that, you can specify an optional `from` and / or `to` parameter using the query string to hand over a timestamp.

If the statistics could be fetched, you get back a `200`.

## Running quality assurance

To run quality assurance for this module use [roboter](https://www.npmjs.com/package/roboter):

```shell
$ npx roboter
```


CHANGELOG.md
# [1.1.0](https://github.com/thenativeweb/rdrct/compare/1.0.2...1.1.0) (2022-03-12)


### Features

* Add statistics route for all urls. ([#20](https://github.com/thenativeweb/rdrct/issues/20)) ([57e0ad1](https://github.com/thenativeweb/rdrct/commit/57e0ad1ee6793612d0604b7e5c6a3c284bd7c4a1))

## [1.0.2](https://github.com/thenativeweb/rdrct/compare/1.0.1...1.0.2) (2022-02-07)


### Bug Fixes

* bump validate-value from 9.2.0 to 9.2.2 ([#15](https://github.com/thenativeweb/rdrct/issues/15)) ([c166477](https://github.com/thenativeweb/rdrct/commit/c166477d970b5e46538b669dcf14c4f3104958a8))

## [1.0.1](https://github.com/thenativeweb/rdrct/compare/1.0.0...1.0.1) (2022-01-25)


### Bug Fixes

* bump defekt from 8.0.0 to 8.1.0 ([#7](https://github.com/thenativeweb/rdrct/issues/7)) ([e4610e6](https://github.com/thenativeweb/rdrct/commit/e4610e68878f2ac931003851a94b9e1d0da2137b))

# 1.0.0 (2022-01-25)


### Bug Fixes

* bump validate-value from 9.1.7 to 9.2.0 ([#9](https://github.com/thenativeweb/rdrct/issues/9)) ([5b8f155](https://github.com/thenativeweb/rdrct/commit/5b8f155361759ade141d4560dbe3ea8af44b1795))


lib/api/addRedirect.ts
import { flaschenpost } from 'flaschenpost';
import { RedirectStore } from '../store/RedirectStore';
import { RequestHandler } from 'express';
import { JsonSchema, Parser } from 'validate-value';

const logger = flaschenpost.getLogger();

const addRedirect = function ({ redirectStore }: {
  redirectStore: RedirectStore;
}): RequestHandler {
  const schemaRequestBody: JsonSchema = {
    type: 'object',
    properties: {
      key: { type: 'string', minLength: 1 },
      url: { type: 'string', minLength: 1, format: 'uri' },
      type: { type: 'string', enum: [ 'permanent', 'temporary' ]}
    },
    required: [ 'key', 'url', 'type' ],
    additionalProperties: false
  };

  const parserRequestBody = new Parser(schemaRequestBody);

  return async function (req, res): Promise<void> {
    try {
      parserRequestBody.parse(req.body).unwrapOrThrow();
    } catch (ex: unknown) {
      logger.warn('Failed to parse request body.', { ex });
      res.status(400).json({ message: (ex as Error).message });

      return;
    }

    const { key, url, type } = req.body;

    let id;

    try {
      id = await redirectStore.add({ key, url, type });
    } catch {
      logger.warn('Redirect already exists.', { key });
      res.status(409).end();

      return;
    }

    logger.info('Added redirect.', { id, key, url, type });
    res.status(200).json({ id });
  };
};

export { addRedirect };
